libnetwork/endpoint.go

```diff
@@ -1178,3 +1159,39 @@ func (c *Controller) cleanupLocalEndpoints() error {
 
 	return nil
 }
+
+// BuildEndpointDNSNames constructs the list of DNSNames that should be assigned to a given endpoint. Any zero-value
+// parameter will be skipped. The order within the returned slice is important as the first entry will be used to
+// generate the PTR records (for IPv4 and v6) associated to this endpoint.
+func BuildEndpointDNSNames(name, hostname, id string, aliases []string) []string {
```

Comment:
This function seems much more related to the daemon business logic than libnetwork: it has specific knowledge about how to transform container names and IDs into appropriate DNS names. And with one exception, it is exclusively used by the daemon to transform container configuration into libnetwork endpoint config. I think this function belongs as an unexported function in the `daemon` package. The one exception, the migration code, only uses a subset of what this function does. The migration code within libnetwork could simply be open-coded. `dedupSlice` could live in some `internal` utility package for use in both. Or we could double down on "live-restore across major upgrades is unsupported" and not bother migrating endpoints at all.